A Program to Detect Recombinants From Unaligned Sequences
-----------------------

### About
This program is novel approach for detecting recombinant sequences and corresponding statistical support values from unaligned sequences. This framework develops on the basis of the paritial alignments from jumping hidden markov model(JHMM), after that, by partitionning them into multiple equal-length triples, on which we uses a novel distance-based procedure to identify recombinant from each triple. Statistical support values calculated from Bootstrap, the bigger, the better, indicating the robustness of identified recombinants.


### Required softwares
- MAFFT used to align one sequence to another two pre_aligned sequences (https://mafft.cbrc.jp/alignment/software/)
- seqkit used to concatenate the two segments for each triple (https://bioinf.shenwei.me/seqkit/download/)
- Python 
-- Require modules for Python 2 user:
-- Require modules for Python 3 user:

### Optional softwares (only used for simulation section)
- Msprime [optional, only used for simulation section] used to generate one arbitrary phylogenetic tree (https://msprime.readthedocs.io/en/stable/installation.html)
- Snakemake
- INDELible
- Python modules  

### Required Input File 
- Patial alignment produced by JHMM (please see MZmosaic folder)
- Input fasta format biological sequences

### Creating Input File
#### Required
- "-i,--input \<filename\>": file path to SV regions in bed file.
- "-o,--output \<filename\>": output file name.
- "-b,--bams \<filename(s)\>": list of bam files corresponding to tumor regions.
- "-f,--frag \<integer(s)\>": comma separated list of average fragment lengths for each bam.
- "-vcf \<filename(s)\>": comma separated list of vcf tardis files with deletions in each bam.
- "-vcfbed \<filename(s)\>": comma separated list of bed files associated with tardis vcfs.

### Run Example

```
python prepareMeltosInput.py -i regions.bed -b PD9770A.bam,PD9770C.bam,PD9770D.bam -vcf PD9770A_del.vcf,PD9770C_del.vcf,PD9770D_del.vcf -vcfbed PD9770A_del.bed,PD9770C_del.bed,PD9770D_del.bed -f 375,353,407 -o regions.meltos
```

### Running Meltos
#### Required:
- "-svFile \<filename\>": name of the file which contains your preprepared SV information.
- "-treeFile \<filename\>": name of the file which contains your premade SSNV lineage tree.
- "-ssnvFile \<filename\>": name of the file which contains your list of SSNVs that were used in your input lineage tree.
- "-numSamples \<integer\>": count of the number of spatially distinct samples in your sv and ssnv files.

#### Optional:
 - "-outputFile \<filename\>": the prefix you want for your outputfiles. If not specified, defaults to the name of your sv file +".meltos"
  NOTE: This will overwrite files with the same name. This prefix should include the directory in which you want the files to go.

 - "-sweep": have meltos run a series of  tests with combinations of different parameters.

 - "-min \<arg\>":   Set an upper bound for an abscence in a sample.

 - "-max \<arg\>":   Set a lower bound for presence in a sample.

 - "-clus \<arg\>":  Set a max difference in VAF allowed for combining variant clusters.

 - "-err \<arg\>":   Set how lenient to be with phylogenic constraints.



### Output
Produces a series of files based on various combinations of parameters, and places them in the directory specified by output.
The file starts with parameter specifications used to create the file.
- Lineage Tree Node Assignments
After that comes VAF calculations and SV assignments for nodes in the lineage tree. Nodes from the original lineage tree will share the same index from the first file, and new nodes will be labeled sequentially after that.
The SV assignments will also share their indices with the indices of the SVs from their original input file. SSNVs are not shown in this output. Node 0 is the root node, and is not from the original lineage tree.
- SVs Assigned to Added New SV Nodes:
A repeat of information from the SV assignments, but only showing nodes that were generated by the algorithm, rather than nodes that originally came from the input tree.
- SVs Assigned to Inviable New SV Nodes:
A list of nodes that were created by the algorithm but were unable to be placed into the lineage tree, together with the indices of their SV assignments.

Finally, the tree's edges are displayed, with the index of the parent node on the left, and the index of the child node on the right.

### Run Example

```
java -jar Meltos.jar -treeFile licheeTestTree.txt -svFile ProcessedSVInput.txt -numSamples 6 -ssnvFile snvVAFFile.txt
```
